pytorch_lightning/profiler/pytorch.py

```diff
@@ -468,23 +380,129 @@ class PyTorchProfiler(LegacyPyTorchProfiler):
                 nvvp trace_name.prof
                 torch.autograd.profiler.load_nvprof(path)
 
-        export_to_flame_graph: warning - Supported only for torch>=1.8.0
+        export_to_flame_graph: warning - Supported only for torch>=1.8
             Whether to export the sequence of profiled operators for Flame Graph.
 
-        on_trace_ready: warning - Supported only for torch>=1.8.0
+        on_trace_ready: warning - Supported only for torch>=1.8
             Function which takes the profiler and executed on ``RECORD_AND_SAVE`` action
 
-        schedule: warning - Supported only for torch>=1.8.0
+        schedule: warning - Supported only for torch>=1.8
             Optional Callable which describes recording procedure
     """
 
 
 if _TORCH_GREATER_EQUAL_1_8:
+    class RegisterRecordFunction:
+        """
+        While profiling autograd operations, this class will add labels
+        with module names around the forward function.
+        The PyTorch Profiler will use this feature automatically.
+
+        It can be deactivated as follows:
+
+        Example::
+
+            from pytorch_lightning.profilers import PyTorchProfiler
+
+            profiler = PyTorchProfiler(record_module_names=False)
+
+            Trainer(profiler=profiler)
+
+        It can be used outside of Lightning as follows:
+
+        Example::
+
+            with RegisterRecordFunction(model):
+                out = model(batch)
+
+        """
+        NN_MODULES: List[str] = dir(torch.nn)
+
+        def __init__(self, model) -> None:
+            self._model = model
+            self._records = {}
+            self._handles = {}
+
+        def _start_recording(self, input: Any, key: Optional[str] = None) -> Any:
+            if key is None:
+                raise ValueError
+            self._records[key] = record_function(key).__enter__()
+            return input
+
+        def _stop_recording(self, result: Any, key: Optional[str] = None) -> Any:
+            if key is None:
+                raise ValueError
+            self._records[key].__exit__(None, None, None)
+            return result
+
+        def __enter__(self):
+            for module_name, module in self._model.named_modules():
+                key = module_name if module in self.NN_MODULES else f"{type(module)}: {module_name}"
+                pre_handle = module.register_forward_pre_hook(partial(self._start_recording, key=key))
+                post_handle = module.register_forward_hook(partial(self._stop_recording, key=key))
+                self._handles[module_name] = [pre_handle, post_handle]
+
+        def __exit__(self, *_):
+            self._records = {}
+            self._handles = {}
```

Comment:
Is this not enough to free up the memory?

Reply:
good question. I am not sure the handles remove the hook if deleted.

Reply:
Lets be safe

```suggestion
            for h in self._handles.values():	
                h.remove()  # remove module forward hooks
            self._records = {}
            self._handles = {}
```